pkg/karmadactl/promote/portforward.go

```diff
@@ -0,0 +1,560 @@
+}
+
+// PortForwarder knows how to listen for local connections and forward them to
+// a remote pod via an upgraded HTTP request.
+type PortForwarder struct {
+	// port forwarding configuration
+	dialer        httpstream.Dialer
+	streamConn    httpstream.Connection
+	Ready         chan struct{}
+	requestIDLock sync.Mutex
+	requestID     int
+	address       string
+	localPort     int
+	remotePort    int
+	listener      io.Closer
+	out           io.Writer
+	errOut        io.Writer
+}
+
+func newWebhookInterpreterHelper(hook configmanager.WebhookAccessor, clientManager *webhookutil.ClientManager, attributes *request.Attributes) *webhookInterpreterHelper {
+	return &webhookInterpreterHelper{
+		hook:          hook,
+		clientManager: clientManager,
+		attributes:    attributes,
+	}
+}
+
+func (wh *webhookInterpreterHelper) callHook(ctx context.Context, context string, kubeconfigPath string, namespace string, pod string, port int) (*request.ResponseAttributes, error) {
+	restConfig, err := apiclient.RestConfig(context, kubeconfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	KubeClientSet, err := apiclient.NewClientSet(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	portforwardReq := KubeClientSet.CoreV1().RESTClient().Post().Namespace(namespace).
+		Resource("pods").Name(pod).SubResource("portforward")
+	ReadyChannel := make(chan struct{})
+
+	res, err := ForwardPorts(ctx, "POST", portforwardReq.URL(), restConfig, ReadyChannel, wh, port)
+	if err != nil {
+		klog.Fatalln(err)
+	}
+
+	return res, nil
+}
+
+func ForwardPorts(ctx context.Context, method string, url *url.URL, config *rest.Config, ReadyChannel chan struct{}, wh *webhookInterpreterHelper, port int) (*request.ResponseAttributes, error) {
+	IOStreams := struct {
+		// In think, os.Stdin
+		In io.Reader
+		// Out think, os.Stdout
+		Out io.Writer
+		// ErrOut think, os.Stderr
+		ErrOut io.Writer
+	}{os.Stdin, os.Stdout, os.Stderr}
+
+	transport, upgrader, err := spdy.RoundTripperFor(config)
+	if err != nil {
+		return nil, err
+	}
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, method, url)
+
+	fw, err := NewOnAddresses(dialer, ReadyChannel, IOStreams.Out, IOStreams.ErrOut, port)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := fw.ForwardPorts(ctx, wh)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+// ForwardPorts formats and executes a port forwarding request.
+func (pf *PortForwarder) ForwardPorts(ctx context.Context, wh *webhookInterpreterHelper) (*request.ResponseAttributes, error) {
+	defer pf.Close()
+
+	var err error
+	pf.streamConn, _, err = pf.dialer.Dial(PortForwardProtocolV1Name)
+	if err != nil {
+		return nil, fmt.Errorf("error upgrading connection: %s", err)
+	}
+	defer pf.streamConn.Close()
+
+	rep, err := pf.forward(ctx, wh)
+	return rep, err
+}
+
+// Close stops the listener of PortForwarder.
+func (pf *PortForwarder) Close() {
+	// stop the listener
+	if err := pf.listener.Close(); err != nil {
```

Comment:
@zhy76 Panic may occur here! If an error is returned on line 291, the close of the listener here is triggered. But because line 295 haven't been reached, the listener has not been initialized yet.